flux_led/__main__.py

```diff
@@ -741,22 +824,56 @@ def setRgbw(self, r,g,b,w, persist=True, brightness=None, retry=2):
         else:
             msg = bytearray([0x41])
 
-        msg.append(int(r))
-        msg.append(int(g))
-        msg.append(int(b))
-        msg.append(int(w))
-        if self.protocol == "LEDENET":
+        if r is not None:
+            msg.append(int(r))
+        else:
+            msg.append(int(0))
+        if g is not None:
+            msg.append(int(g))
+        else:
+            msg.append(int(0))
+        if b is not None:
+            msg.append(int(b))
+        else:
+            msg.append(int(0))
+        if w is not None:
             msg.append(int(w))
+        else:
+            msg.append(int(0))
+
+        # LEDENET devices support two white outputs for cool and warm. We set
+        # the second one here - if we're only setting a single white value,
+        # we set the second output to be the same as the first
+        if self.protocol == "LEDENET":
+            if w2 is not None:
+                msg.append(int(w2))
+            elif w is not None:
+                msg.append(int(w))
+            else:
+                msg.append(0)
 
         if not self.rgbwprotocol:
-            if w > 0:
+            # For devices that can't set RGB+W simultaneously, indicate whether
+            # we should set the white outputs or the RGB outputs
+            if w is not None or w2 is not None:
                 msg.append(0x0f)
             else:
                 msg.append(0xf0)
         else:
+            # These devices can simultaneously control the RGB and white output
             msg.append(0x00)
 
+        # Message terminator
         msg.append(0x0f)
+
+        # yet another type of protocol/controller that only supports RGB (no ww)
+        if self.protocol == 'LEDnet_wifi370':
+            msg = bytearray([0x56])
+            msg.append(int(r))
+            msg.append(int(g))
+            msg.append(int(b))
+            msg.append(0xaa)
+        
```

Comment:
You are right. there should be a better way to do this. As I see it, we have 4 different cases  we need to support: Original Ledenet, LEDENET, badrgb and other devices. I set out differences below:

`        # sample message for original LEDENET protocol
        #  0  1  2  3  4
        # 56 90 fa 77 aa
        #  |  |  |  |  |
        #  |  |  |  |  terminator
        #  |  |  |  blue
        #  |  |  green
        #  |  red
        #  head
        #
        # sample message for the other protocols
        #  0  1  2  3  4  5  6
        # 31 90 fa 77 00 00 0f
        #  |  |  |  |  |  |  |
        #  |  |  |  |  |  |  terminator
        #  |  |  |  |  |  special/white2 (see below)
        #  |  |  |  |  white
        #  |  |  |  blue
        #  |  |  green
        #  |  red
        #  persisitence (31 for true / 41 for false)
        #
        # the special byte
        # the special byte can have different values depending on the type
        # of device:
        # For devices that support 2 types of white value (warm and cold
        # white) this value is the cold white value. These use the LEDENET
        # protocol. If a second value is not given, reuse the first white value.
        #
        # For devices that need to set rbgw in two commands, this value
        # specifies if this command is to set white value (0f) or the rgb
        # value (f0). these devices have the protocol BadRGBW
        #
        # For all other rgb and rgbw devices, the value is 00
`